### PR TITLE
added validation of the HTTP response - FE

### PR DESCRIPTION
**Pull request by yaelmualem**
added validation of the HTTP response. in case of an error, when the status code isn't 2xx, an error message pops out :
image

### CODE REVIEW

**Comment by yaelmualem** [comment]
@ScDor FYI

File: Closure_Front_End/src/views/Home.vue [hunk 1]
```diff
@@ -158,9 +159,12 @@ export default {
   created() {
     if (this.$auth.isAuthenticated.value) {
       this.$auth.getIdTokenClaims().then(console.log, console.error);
-      this.$http.get("tracks/?track_number=3010").then(response => {
+      this.$http.get("tracks/?track_number=3010",
+      {validateStatus: Utils.validateStatusCode(status)}
+      ).then(response => {
         this.track = response.data.results[0]
       })
+      .catch (Utils.popOutStatusCodeError);
```
[the remark this answers is not in the document]
Comment (yaelmualem) [reply Closure_Front_End/src/views/Home.vue:167]: this isn't my change:
https://github.com/ScDor/Closure/blob/master/Closure_Front_End/src/views/Home.vue#:~:text=%7D%2C-,created()%20%7B,%7D,-%7D%2C
this is the default home page (main page) I guess that @noabarlia added it with default values since it is easier to develop with the hardcoded values (there is more default hardcoded values above). It will need to be removed before the site is aired, but for now, it is ok. @noabarlia correct me if I'm wrong.

File: Closure_Front_End/src/components/SearchBar.vue [hunk 2]
```diff
@@ -48,9 +51,16 @@ export default {
    * more on the implementation of the search on the backend part of the project.
    */ {
     search() {
-      this.$http.get(this.url + this.query)
+      axios
+        .get(this.url + this.query, {
+          headers: {
+            Authorization: "Token 425fa39de10f02351c7043d0dbe34a4b31be7a27",
```
[the remark this answers is not in the document]
Comment (yaelmualem) [reply Closure_Front_End/src/components/SearchBar.vue]: fix, thanks!

File: Closure_Front_End/src/views/Home.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (ScDor) [reply Closure_Front_End/src/views/Home.vue:167]: Thanks, opened #93

**Comment by rbenjos** [comment]
looks great, but please try and not use axios in SearchBar.vue, do as you did in the Home view

**Review by rbenjos — CHANGES_REQUESTED** [review]
looks great, but please try and not use axios in SearchBar.vue, do as you did in the Home view

File: Closure_Front_End/src/Utils.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ */
+function validateStatusCode (statusCode) {
+    return statusCode >= 200 && statusCode < 400
+}
+
+/**
+ * pops out an alert with generic error message. We call it when there was an error while performing an HTTP call.
+ */
+function popOutStatusCodeError(){
+    alert("Error occurred. Couldn't perform the operation.")
```
Comment (nmdanny) [inline Closure_Front_End/src/Utils.js:14]: Alerts aren't a very good way of showing errors:

- We should use a proper Vue component for displaying a "toast", which could be used for many kinds of errors, not just network related errors. Something that looks like one of the error notifications [here](https://bulma.io/documentation/elements/notification/)

- The error message should be informative(and in Hebrew, at least some of it)
  For example, error codes that begin with 500 indicate a problem in our server's code
  Error code 404 might indicate that the server is down
  A network error might happen if the client's internet is down(in that case, there wouldn't be any HTTP status code)

File: Closure_Front_End/src/components/SearchBar.vue [hunk 4]
```diff
@@ -48,9 +50,13 @@ export default {
...
     search() {
-      this.$http.get(this.url + this.query)
+      this.$http
+        .get(this.url + this.query, {
+          validateStatus: Utils.validateStatusCode(status)
+        })
         .then((response) => (this.suggestions = response.data.results))
-        .then((this.hide = !this.suggestions.length || this.query == ""));
+        .then((this.hide = !this.suggestions.length || this.query == ""))
+        .catch (Utils.popOutStatusCodeError);
```
Comment (nmdanny) [inline Closure_Front_End/src/components/SearchBar.vue:59]: Manually adding a 'catch' and 'validateStatus' to every `$http` call leads to a lot of duplication

I recommend putting all axios related error handling in a single point, which should also be tied to a vue component for error handling.
See [this](https://stackoverflow.com/questions/48990632/how-to-manage-axios-errors-globally-or-from-one-point) - an example of using an [interceptor](https://axios-http.com/docs/interceptors) to automatically handle errors. This can be done to the axios instance created at 
https://github.com/ScDor/Closure/blob/a48ac99eb8d9958dce73c84bbeb373798c9e538c/Closure_Front_End/src/auth/index.js#L108

(Similarly to how I set the base URL, one can also set `validateStatus` as explained [here](https://axios-http.com/docs/req_config))

Comment (nmdanny) [reply Closure_Front_End/src/components/SearchBar.vue:59]: To clarify, there may be cases where we want to use a `.catch` to deal with specific types of errors that should be handled differently. For example, a business logic error (e.g, adding a non existent course, conflicting semester, etc..) should present a much more informative error UI, that could for example highlight the problematic course, this is in contrast to generic network errors or internal errors that would be dealt by an showing a toast

File: Closure_Front_End/src/Utils.js (hunk 3 above)
Comment (yaelmualem) [reply Closure_Front_End/src/Utils.js:14]: why alert isn't good?

Comment (nmdanny) [reply Closure_Front_End/src/Utils.js:14]: `console.alert` doesn't provide a good user experience, see [here](https://stackoverflow.com/questions/8825384/alert-is-bad-really) and [here](https://www.quora.com/Why-is-JavaScripts-alert-box-considered-a-bad-design-from-usability-point-of-view)

For example, suppose you're trying to search for courses and there are network errors. With every letter you type, you'll get an alert box

What about errors that don't deserve such a disruptive alert box? For example, maybe we'll implement caching of  courses/tracks within local-storage, so if the network requests fail, we'll get a more limited selection of courses, but we should still be able to proceed with the app

Or, how about auto-save: our app could auto-save the User courses into both local storage and to the DB. Without network access, the user should get a warning that his changes aren't saved to the DB but are still saved on the local-storage of his browser. 

Basically, in many scenarios where the user has no ways of dealing with the error(e.g, network errors), it is best to warn him about it in a non-intrusive manner (e.g, a "toast")

And in the cases where he can deal with it(business logic errors), then we'll probably want a nicer and more informative way of letting him deal with it.

Comment (yaelmualem) [reply Closure_Front_End/src/Utils.js:14]: got it, thanks :) I will fix it

**Comment by yaelmualem** [comment]
@nmdanny @rbenjos entered the changes that you requested. can you look at the PR again?
Thanks @nmdanny for your help!

File: Closure_Front_End/src/views/Home.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (noabarlia) [reply Closure_Front_End/src/views/Home.vue]: Yes, it's there for development purposes

**Comment by nmdanny** [comment]
Looks great!

A minor nitpick - I think the error box looks better without a `min-height`, otherwise there's too much wasted space for short error messages

**Review by rbenjos — APPROVED** [review]
